File: ios-tests/scripts/generate-testids.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "pathname"

REPO_ROOT = Pathname.new(File.expand_path("../..", __dir__))
SOURCE = REPO_ROOT.join("app/src/constants/testIds.ts")
OUTPUT = REPO_ROOT.join("ios-tests/MobileTaskManagerUITests/TestIds.swift")

contents = SOURCE.read
block = contents[/export const TestIds = \{(.+?)\} as const;/m, 1]
raise "Could not locate the TestIds object in #{SOURCE}" unless block

entries = block.scan(/^\s*([A-Za-z0-9_]+):\s*'([^']*)',?\s*$/)
raise "No testID entries found in #{SOURCE}" if entries.empty?

lines = entries.map { |key, value| "  static let #{key} = \"#{value}\"" }

functions = contents.scan(
  /export function (\w+)\((\w+): (number|string)\): string \{\s*return `([^`]+)`;\s*\}/,
)
function_lines = functions.map do |name, param, ts_type, template|
  swift_type = ts_type == "number" ? "Int" : "String"
  swift_template = template.gsub(/\$\{#{param}\}/) { "\\(#{param})" }
  "  static func #{name}(_ #{param}: #{swift_type}) -> String { \"#{swift_template}\" }"
end

body = lines
body += ["", *function_lines] unless function_lines.empty?

OUTPUT.write(<<~SWIFT)
  enum TestIds {
  #{body.join("\n")}
  }
SWIFT

puts "Generated #{OUTPUT.relative_path_from(REPO_ROOT)} " \
     "(#{entries.length} identifiers, #{functions.length} helpers)."

File: android-tests/scripts/generate-testids.mjs
#!/usr/bin/env node

import { readFileSync, writeFileSync } from 'node:fs';
import { dirname, join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';

const REPO_ROOT = join(dirname(fileURLToPath(import.meta.url)), '..', '..');
const SOURCE = join(REPO_ROOT, 'app/src/constants/testIds.ts');
const OUTPUT = join(
  REPO_ROOT,
  'android-tests/src/androidTest/kotlin/com/mobiletaskmanager/TestIds.kt',
);

const KOTLIN_TYPES = { number: 'Int', string: 'String' };

const contents = readFileSync(SOURCE, 'utf8');

const block = contents.match(
  /export const TestIds = \{(.+?)\} as const;/s,
)?.[1];
if (!block) {
  throw new Error(`Could not locate the TestIds object in ${SOURCE}`);
}

const entries = [...block.matchAll(/^\s*([A-Za-z0-9_]+):\s*'([^']*)',?\s*$/gm)];
if (entries.length === 0) {
  throw new Error(`No testID entries found in ${SOURCE}`);
}

const constants = entries.map(
  ([, key, value]) => `  const val ${key} = "${value}"`,
);

const functions = [
  ...contents.matchAll(
    /export function (\w+)\((\w+): (number|string)\): string \{\s*return `([^`]+)`;\s*\}/g,
  ),
];

const functionLines = functions.map(([, name, param, tsType, template]) => {
  const kotlinType = KOTLIN_TYPES[tsType];
  const kotlinTemplate = template.replaceAll(`\${${param}}`, `$${param}`);
  return `  fun ${name}(${param}: ${kotlinType}): String = "${kotlinTemplate}"`;
});

const body = [...constants];
if (functionLines.length > 0) {
  body.push('', ...functionLines);
}

writeFileSync(
  OUTPUT,
  `package com.mobiletaskmanager

object TestIds {
${body.join('\n')}
}
`,
);

console.log(
  `Generated ${relative(REPO_ROOT, OUTPUT)} ` +
    `(${entries.length} identifiers, ${functions.length} helpers).`,
);
